### ISSUE TITLE
nobody2018 - BalancerPairOracle#getPrice will revert due to division by zero in some cases

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-05-blueberry-judging#25]
nobody2018

medium

# BalancerPairOracle#getPrice will revert due to division by zero in some cases

## Summary

`BalancerPairOracle#getPrice` internally calls `computeFairReserves`, which returns fair reserve amounts given spot reserves, weights, and fair prices. When the parameter `resA` passed to `computeFairReserves` is smaller than `resB`, division by 0 will occur.

## Vulnerability Detail

In `BalancerPairOracle#getPrice`, resA and resB passed to `computeFairReserves` are the balance of TokenA and TokenB of the pool respectively. It is common for the balance of TokenB to be greater than the balance of TokenA.

```solidity
function computeFairReserves(
        uint256 resA,
        uint256 resB,
        uint256 wA,
        uint256 wB,
        uint256 pxA,
        uint256 pxB
    ) internal pure returns (uint256 fairResA, uint256 fairResB) {
    	...
    	//@audit r0 = 0 when resA < resB.
->      uint256 r0 = resA / resB;
        uint256 r1 = (wA * pxB) / (wB * pxA);
        // fairResA = resA * (r1 / r0) ^ wB
        // fairResB = resB * (r0 / r1) ^ wA
        if (r0 > r1) {
            uint256 ratio = r1 / r0;
            fairResA = resA * (ratio ** wB);
            fairResB = resB / (ratio ** wA);
        } else {
->          uint256 ratio = r0 / r1;		// radio = 0 when r0 = 0
->          fairResA = resA / (ratio ** wB);   	// revert divided by 0
            fairResB = resB * (ratio ** wA);
        }
    }
```

Another case is **when the decimals of tokenA is smaller than the decimals of tokenB**, such as usdc(e6)-weth(e18).

## Impact

All functions that subcall `BalancerPairOracle#getPrice` will be affected.

## Code Snippet

https://github.com/sherlock-audit/2023-05-blueberry/blob/main/blueberry-core/contracts/oracle/BalancerPairOracle.sol#L35-L66

## Tool used

Manual Review

## Recommendation

```diff
--- a/blueberry-core/contracts/oracle/BalancerPairOracle.sol
+++ b/blueberry-core/contracts/oracle/BalancerPairOracle.sol
@@ -50,7 +50,7 @@ contract BalancerPairOracle is UsingBaseOracle, IBaseOracle {
         // --> fairResA / r1^wB = constant product
         // --> fairResA = resA^wA * resB^wB * r1^wB
         // --> fairResA = resA * (resB/resA)^wB * r1^wB = resA * (r1/r0)^wB
-        uint256 r0 = resA / resB;
+        uint256 r0 = resA * 10**(decimalsB) / resB;
         uint256 r1 = (wA * pxB) / (wB * pxA);
         // fairResA = resA * (r1 / r0) ^ wB
         // fairResB = resB * (r0 / r1) ^ wA
```

**Comment by securitygrid** [comment]
Escalate for 10 USDC
This is a valid M/H. Please review **Vulnerability Detail** of this report, it describes 2 cases:
1. resA and resB passed to computeFairReserves are the balance of TokenA and TokenB of the pool respectively. It is common for the balance of TokenB to be greater than the balance of TokenA.
2. when the decimals of tokenA is smaller than the decimals of tokenB, such as usdc(e6)-weth(e18).

**This issue is same root as #28**.
The impact described in #28 is only possible if the dp of the two tokens are different. But it is not completely correct, because when the dp of token0 is less than the dp of token1, a division by zero error occurs, such as token0(e6)-token1(e18).
Merging the two reports is the best description.

**Comment by sherlock-admin** [comment]
> Escalate for 10 USDC
> This is a valid M/H. Please review **Vulnerability Detail** of this report, it describes 2 cases:
> 1. resA and resB passed to computeFairReserves are the balance of TokenA and TokenB of the pool respectively. It is common for the balance of TokenB to be greater than the balance of TokenA.
> 2. when the decimals of tokenA is smaller than the decimals of tokenB, such as usdc(e6)-weth(e18).
> 
> **This issue is same root as #28**.
> The impact described in #28 is only possible if the dp of the two tokens are different. But it is not completely correct, because when the dp of token0 is less than the dp of token1, a division by zero error occurs, such as token0(e6)-token1(e18).
> Merging the two reports is the best description.

You've created a valid escalation for 10 USDC!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by IAm0x52** [comment]
Agreed with escalation. This and #28 are dupes and this does a better job of describing the issue so it should be the main issue. Additionally given that the contract would become nonfunctional rather than return an incorrect price, I agree with the watson's original severity of medium.

**Comment by hrishibhat** [comment]
Escalation accepted

Valid medium
Making this issue the main one and #28 a duplicate of this issue.

**Comment by sherlock-admin** [comment]
> Escalation accepted
> 
> Valid medium
> Making this issue the main one and #28 a duplicate of this issue. 
> 

    This issue's escalations have been accepted!

    Contestants' payouts and scores will be updated according to the changes made on this issue.

**Comment by Gornutz** [comment]
https://github.com/Blueberryfi/blueberry-core/commit/e5348bc649822f6e96c9c7be25d2b6a06d5a6318#diff-d8a6f8d601164fa5805ceb20da1f406bef6792d6f33fa002c258d1f63ccb25b3

**Comment by IAm0x52** [comment]
This adds an outdated library (BNum.sol). Needs a different math library.
